Classification/FeatureSelection/src/main/java/org/tribuo/classification/fs/FS_Wrapper_Approaches/Optimizers/CuckooSearchOptimizer.java

```diff
@@ -0,0 +1,157 @@
+package FS_Wrapper_Approaches.Optimizers;
+
+import FS_Wrapper_Approaches.Discreeting.Binarizing;
+import FS_Wrapper_Approaches.Discreeting.TransferFunction;
+import com.oracle.labs.mlrg.olcut.util.Pair;
+import org.tribuo.*;
+import org.tribuo.classification.Label;
+import org.tribuo.classification.evaluation.LabelEvaluation;
+import org.tribuo.classification.evaluation.LabelEvaluator;
+import org.tribuo.common.nearest.KNNClassifierOptions;
+import org.tribuo.dataset.SelectedFeatureDataset;
+import org.tribuo.evaluation.CrossValidation;
+import org.tribuo.provenance.FeatureSelectorProvenance;
+import org.tribuo.provenance.FeatureSetProvenance;
+import org.tribuo.provenance.impl.FeatureSelectorProvenanceImpl;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+
+/**
+ * Select features based on Cuckoo Search algorithm with binary transfer functions, KNN classifier and 10-fold cross validation
+ * <p>
+ * see:
+ * <pre>
+ * Xin-She Yang and Suash Deb.
+ * "Cuckoo Search via L´evy Flights", 2010.
+ *
+ * L. A. M. Pereira et al.
+ * "A Binary Cuckoo Search and its Application for Feature Selection", 2014.
+ * </pre>
+ */
+public class CuckooSearchOptimizer implements FeatureSelector<Label> {
+    private final TransferFunction transferFunction;
+    private final double stepSizeScaling;
+    private final double lambda;
+    private final double worstNestProbability;
+    private final double delta;
+    private final int populationSize;
+    private int [][] setOfSolutions;
+    private final int maxIteration;

```

Comment:
All the final variables here should not be `final` and need to be tagged `@Config` (with appropriate description fields) so they are automatically captured by the provenance system.